frontend/expr_inferrer.py

```diff
@@ -0,0 +1,360 @@
+"""Inferrer for python expressions.
+
+Infers the types for the following expressions:
+     - BinOp(expr left, operator op, expr right)
+     - UnaryOp(unaryop op, expr operand)
+     - Dict(expr* keys, expr* values)
+     - Set(expr* elts)
+     - Num(object n)
+     - Str(string s)
+     - NameConstant(singleton value)
+     - List(expr* elts, expr_context ctx)
+     - Tuple(expr* elts, expr_context ctx)
+     - Bytes(bytes s)
+     - IfExp(expr test, expr body, expr orelse)
+	 - Subscript(expr value, slice slice, expr_context ctx)
+	 - Await(expr value) --> Python 3.5+
+	 - Yield(expr? value)
+     - Compare(expr left, cmpop* ops, expr* comparators)
+     - Name(identifier id, expr_context ctx)
+     - FormattedValue(expr value, int? conversion, expr? format_spec) --> Python 3.6+
+     - JoinedStr(expr* values) --> Python 3.6+
+     - ListComp(expr elt, comprehension* generators)
+     - SetComp(expr elt, comprehension* generators)
+     - DictComp(expr key, expr value, comprehension* generators)
+
+     TODO:
+     - Lambda(arguments args, expr body)
+     - GeneratorExp(expr elt, comprehension* generators)
+     - YieldFrom(expr value)
+     - Call(expr func, expr* args, keyword* keywords)
+     - Attribute(expr value, identifier attr, expr_co0ontext ctx)
+     - Starred(expr value, expr_context ctx)
+
+     TODO:
+     - Infer (or narrow) types based on context. For example, in the following expression:
+        (x + 2) * 3
+        if the type of x is still not inferred (or inferred to be a union type), we can infer (narrow) that x
+        is a numeric value (TBool, TInt or TFloat)
+        Affected inference functions:
+            - infer_binary_operation
+            - infer_unary_operation
+            - infer_subscript
+            - infer_compare
+"""
+
+import ast, sys
+from i_types import *
+from context import Context
+from abc import ABCMeta, abstractmethod
+from exceptions import HomogeneousTypesConflict
+
+def infer_numeric(node):
+    """Infer the type of a numeric node"""
+    if type(node.n) == int:
+        return TInt()
+    if type(node.n) == float:
+        return TFloat()
+
+def __get_supertype(elts, context):
+    if len(elts) == 0:
+        return TNone()
```

Comment:
I think we need the type that comes at the leaves of subtype hierarchy (not the root), which will be a subtype of any other type. What do you think?